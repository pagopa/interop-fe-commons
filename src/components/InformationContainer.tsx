import React from 'react'
import { Box, Typography, Stack, Divider, Skeleton } from '@mui/material'
import type { StackProps } from '@mui/material'
import { CopyToClipboardButton } from '@pagopa/mui-italia'
import type { CopyToClipboardProps } from '@pagopa/mui-italia'

export interface InformationContainerProps extends Omit<StackProps, 'children' | 'content'> {
  label: string
  labelDescription?: string
  content: JSX.Element | string
  copyToClipboard?: CopyToClipboardProps
}

/**
 * Renders a label and a content in a row or a column.
 *
 * If the content is a string, it will be wrapped in a <p> tag.
 * If the content is a JSX.Element, it will be wrapped in a <div> tag.
 *
 * If the copyToClipboard prop is passed, a copy to clipboard button will be rendered.
 * The copyToClipboard prop will be passed to the CopyToClipboardButton `@pagopa/mui-italia's` component.
 */
export function InformationContainer({
  label,
  labelDescription,
  copyToClipboard,
  direction = 'row',
  content,
  ...props
}: InformationContainerProps) {
  return (
    <Stack spacing={direction === 'column' ? 0 : 4} direction={direction} {...props}>
      <Box sx={{ flexShrink: 0, maxWidth: direction === 'column' ? 'none' : '200px', flex: 1 }}>
        <Typography sx={{ mt: copyToClipboard ? 1 : undefined }} variant="body2">
          {label}
        </Typography>
        {labelDescription && (
          <>
            <Divider sx={{ my: 0.5 }} />
            <Typography color="text.secondary" sx={{ display: 'inline-block' }} variant="caption">
              {labelDescription}
            </Typography>
          </>
        )}
      </Box>
      <Box sx={{ flex: 1 }}>
        <Typography
          sx={{ display: 'flex', alignItems: 'start' }}
          component={typeof content === 'string' ? 'p' : 'div'}
          variant="body2"
          fontWeight={600}
        >
          <Box component="span" sx={{ display: 'block', mt: copyToClipboard ? 1 : undefined }}>
            {content}
          </Box>
          {copyToClipboard && <CopyToClipboardButton {...copyToClipboard} />}
        </Typography>
      </Box>
    </Stack>
  )
}

/**
 * Renders a skeleton for the InformationContainer component.
 */
export const InformationContainerSkeleton: React.FC = () => {
  return (
    <Stack spacing={4} direction="row">
      <Box sx={{ flexShrink: 0, maxWidth: '200px', flex: 1 }}>
        <Skeleton />
      </Box>
      <Box sx={{ flex: 1 }}>
        <Skeleton />
        <Skeleton />
      </Box>
    </Stack>
  )
}
