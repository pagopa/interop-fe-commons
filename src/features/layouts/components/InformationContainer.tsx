import React from 'react'
import { Box, Typography, Stack, Divider, Skeleton } from '@mui/material'
import type { StackProps } from '@mui/material'
import { CopyToClipboardButton } from '@pagopa/mui-italia'
import type { CopyToClipboardProps } from '@pagopa/mui-italia'

export interface InformationContainerProps extends Omit<StackProps, 'children'> {
  label: string
  labelDescription?: string
  content: JSX.Element | string
  copyToClipboard?: CopyToClipboardProps
}

export function InformationContainer({
  label,
  labelDescription,
  copyToClipboard,
  direction = 'row',
  content,
  ...props
}: InformationContainerProps) {
  return (
    <Stack spacing={direction === 'column' ? 0 : 4} direction={direction} {...props}>
      <Box sx={{ flexShrink: 0, maxWidth: direction === 'column' ? 'none' : '200px', flex: 1 }}>
        <Typography variant="body2">{label}</Typography>
        {labelDescription && (
          <>
            <Divider sx={{ my: 0.5 }} />
            <Typography color="text.secondary" sx={{ display: 'inline-block' }} variant="caption">
              {labelDescription}
            </Typography>
          </>
        )}
      </Box>
      <Box sx={{ flex: 1 }}>
        <Typography
          component={typeof content === 'string' ? 'p' : 'div'}
          variant="body2"
          fontWeight={600}
        >
          {content}
          {copyToClipboard && <CopyToClipboardButton {...copyToClipboard} />}
        </Typography>
      </Box>
    </Stack>
  )
}

export const InformationContainerSkeleton: React.FC = () => {
  return (
    <Stack spacing={4} direction="row">
      <Box sx={{ flexShrink: 0, maxWidth: '200px', flex: 1 }}>
        <Skeleton />
      </Box>
      <Box sx={{ flex: 1 }}>
        <Skeleton />
        <Skeleton />
      </Box>
    </Stack>
  )
}
