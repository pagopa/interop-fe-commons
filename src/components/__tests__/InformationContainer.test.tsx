import React from 'react'
import { InformationContainer, InformationContainerSkeleton } from '../InformationContainer'
import renderer from 'react-test-renderer'

describe('InformationContainer component', () => {
  it('Renders correctly', () => {
    const tree = renderer
      .create(<InformationContainer label={'test-title'} content="information" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Renders the loading skeleton correctly', () => {
    const tree = renderer.create(<InformationContainerSkeleton />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Render correctly with column direction', () => {
    const tree = renderer
      .create(
        <InformationContainer direction="column" label={'test-title'} content="information" />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Render correctly with content as tring', () => {
    const tree = renderer
      .create(
        <InformationContainer
          label={'test-title'}
          labelDescription="label-description"
          content="information"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Render correctly with JSX as content', () => {
    const tree = renderer
      .create(
        <InformationContainer
          label={'test-title'}
          labelDescription="label-description"
          content={<p>information</p>}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Render correctly with copy to clipboard button', () => {
    const tree = renderer
      .create(
        <InformationContainer
          label={'test-title'}
          copyToClipboard={{ value: 'test', tooltipTitle: 'test' }}
          content="information"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
