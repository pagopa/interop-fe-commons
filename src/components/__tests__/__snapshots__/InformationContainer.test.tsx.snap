// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InformationContainer component > Render correctly with JSX as content 1`] = `
<div
  className="MuiStack-root css-1xcvc1q-MuiStack-root"
>
  <div
    className="MuiBox-root css-iuqs4u"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      test-title
    </p>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-159wgzx-MuiDivider-root"
    />
    <span
      className="MuiTypography-root MuiTypography-caption css-g171c3-MuiTypography-root"
    >
      label-description
    </span>
  </div>
  <div
    className="MuiBox-root css-1rr4qq7"
  >
    <div
      className="MuiTypography-root MuiTypography-body2 css-1vn33h7-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-13o7eu2"
      >
        <p>
          information
        </p>
      </span>
    </div>
  </div>
</div>
`;

exports[`InformationContainer component > Render correctly with column direction 1`] = `
<div
  className="MuiStack-root css-nen11g-MuiStack-root"
>
  <div
    className="MuiBox-root css-1pdb85g"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      test-title
    </p>
  </div>
  <div
    className="MuiBox-root css-1rr4qq7"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-1vn33h7-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-13o7eu2"
      >
        information
      </span>
    </p>
  </div>
</div>
`;

exports[`InformationContainer component > Render correctly with content as tring 1`] = `
<div
  className="MuiStack-root css-1xcvc1q-MuiStack-root"
>
  <div
    className="MuiBox-root css-iuqs4u"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      test-title
    </p>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-159wgzx-MuiDivider-root"
    />
    <span
      className="MuiTypography-root MuiTypography-caption css-g171c3-MuiTypography-root"
    >
      label-description
    </span>
  </div>
  <div
    className="MuiBox-root css-1rr4qq7"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-1vn33h7-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-13o7eu2"
      >
        information
      </span>
    </p>
  </div>
</div>
`;

exports[`InformationContainer component > Render correctly with copy to clipboard button 1`] = `
<div
  className="MuiStack-root css-1xcvc1q-MuiStack-root"
>
  <div
    className="MuiBox-root css-iuqs4u"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-1xwhmis-MuiTypography-root"
    >
      test-title
    </p>
  </div>
  <div
    className="MuiBox-root css-1rr4qq7"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-1vn33h7-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-k1s5lc"
      >
        information
      </span>
      <button
        aria-label="Copia"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-vqczhe-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="ContentCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </p>
  </div>
</div>
`;

exports[`InformationContainer component > Renders correctly 1`] = `
<div
  className="MuiStack-root css-1xcvc1q-MuiStack-root"
>
  <div
    className="MuiBox-root css-iuqs4u"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      test-title
    </p>
  </div>
  <div
    className="MuiBox-root css-1rr4qq7"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-1vn33h7-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-13o7eu2"
      >
        information
      </span>
    </p>
  </div>
</div>
`;

exports[`InformationContainer component > Renders the loading skeleton correctly 1`] = `
<div
  className="MuiStack-root css-1xcvc1q-MuiStack-root"
>
  <div
    className="MuiBox-root css-iuqs4u"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
      style={
        {
          "height": undefined,
          "width": undefined,
        }
      }
    />
  </div>
  <div
    className="MuiBox-root css-1rr4qq7"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
      style={
        {
          "height": undefined,
          "width": undefined,
        }
      }
    />
    <span
      className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
      style={
        {
          "height": undefined,
          "width": undefined,
        }
      }
    />
  </div>
</div>
`;
